File: backend/app/schemas/vehicle.py
"""
車両データスキーマ
TDD Green Phase - テストを通すための最小実装
API_DESIGN.md準拠
"""

from typing import Optional
from decimal import Decimal
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict


class StoreInfo(BaseModel):
    """車両レスポンス用の店舗情報（簡易版）"""

    id: str = Field(..., description="店舗ID")
    name: str = Field(..., description="店舗名")
    address: str = Field(..., description="住所")

    model_config = ConfigDict(from_attributes=True)


class VehicleCreate(BaseModel):
    """車両作成用スキーマ"""

    # 基本情報
    make: str = Field(..., max_length=50, description="メーカー")
    model: str = Field(..., max_length=50, description="モデル名")
    year: int = Field(..., ge=1900, le=2030, description="年式")
    color: str = Field(..., max_length=30, description="色")
    license_plate: str = Field(..., max_length=20, description="ナンバープレート")

    # 分類
    category: str = Field(..., max_length=20, description="カテゴリ")
    class_type: str = Field(..., max_length=20, description="クラス分類")

    # 仕様
    transmission: str = Field(..., max_length=10, description="トランスミッション")
    fuel_type: str = Field(..., max_length=15, description="燃料タイプ")

    # 料金
    daily_rate: Decimal = Field(..., gt=0, description="日額料金")

    # 所在店舗
    store_id: Optional[str] = Field(None, description="所在店舗ID")

    # 画像
    image_filename: Optional[str] = Field(
        None, max_length=100, description="画像ファイル名"
    )
    
    # 喫煙可否
    is_smoking_allowed: Optional[bool] = Field(False, description="喫煙可否")


class VehicleResponse(BaseModel):
    """車両レスポンス用スキーマ"""

    # ID
    id: str = Field(..., description="車両ID")

    # 基本情報
    make: str = Field(..., description="メーカー")
    model: str = Field(..., description="モデル名")
    year: int = Field(..., description="年式")
    color: str = Field(..., description="色")
    license_plate: str = Field(..., description="ナンバープレート")

    # 分類
    category: str = Field(..., description="カテゴリ")
    class_type: str = Field(..., description="クラス分類")
    transmission: str = Field(..., description="トランスミッション")
    fuel_type: str = Field(..., description="燃料タイプ")

    # 料金
    daily_rate: Decimal = Field(..., description="日額料金")

    # ステータス
    is_available: bool = Field(..., description="利用可能フラグ")
    
    # 喫煙可否
    is_smoking_allowed: bool = Field(..., description="喫煙可否")

    # 画像
    image_filename: Optional[str] = Field(None, description="画像ファイル名")

    # 所在店舗
    store_id: Optional[str] = Field(None, description="所在店舗ID")

    # 店舗情報
    store: Optional[StoreInfo] = Field(None, description="所在店舗情報")

    # タイムスタンプ
    created_at: datetime = Field(..., description="作成日時")
    updated_at: datetime = Field(..., description="更新日時")

    model_config = ConfigDict(from_attributes=True)  # SQLAlchemyモデルからの変換を有効化


class VehicleUpdate(BaseModel):
    """車両更新用スキーマ（管理者用）"""

    # 基本情報（オプショナル）
    make: Optional[str] = Field(None, max_length=50, description="メーカー")
    model: Optional[str] = Field(None, max_length=50, description="モデル名")
    year: Optional[int] = Field(None, ge=1900, le=2030, description="年式")
    color: Optional[str] = Field(None, max_length=30, description="色")

    # 分類
    category: Optional[str] = Field(None, max_length=20, description="カテゴリ")
    class_type: Optional[str] = Field(None, max_length=20, description="クラス分類")

    # 仕様
    transmission: Optional[str] = Field(None, max_length=10, description="トランスミッション")
    fuel_type: Optional[str] = Field(None, max_length=15, description="燃料タイプ")

    # 料金
    daily_rate: Optional[Decimal] = Field(None, gt=0, description="日額料金")

    # ステータス（管理者が更新可能）
    is_available: Optional[bool] = Field(None, description="利用可能フラグ")
    is_active: Optional[bool] = Field(None, description="アクティブフラグ")
    
    # 喫煙可否
    is_smoking_allowed: Optional[bool] = Field(None, description="喫煙可否")

    # 所在店舗
    store_id: Optional[str] = Field(None, description="所在店舗ID")

    # 画像
    image_filename: Optional[str] = Field(
        None, max_length=100, description="画像ファイル名"
    )


class VehicleListResponse(BaseModel):
    """車両一覧表示用の軽量スキーマ"""

    id: str = Field(..., description="車両ID")
    make: str = Field(..., description="メーカー")
    model: str = Field(..., description="モデル名")
    year: int = Field(..., description="年式")
    category: str = Field(..., description="カテゴリ")
    daily_rate: Decimal = Field(..., description="日額料金")
    is_available: bool = Field(..., description="利用可能フラグ")
    image_filename: Optional[str] = Field(None, description="画像ファイル名")
    transmission: str = Field(..., description="トランスミッション")
    is_smoking_allowed: bool = Field(..., description="喫煙可否")

    # 店舗情報
    store: Optional[StoreInfo] = Field(None, description="所在店舗情報")

    model_config = ConfigDict(from_attributes=True)

File: frontend/lib/types/vehicle.ts
/**
 * Vehicle関連の型定義
 * バックエンドのPydanticスキーマと対応
 */

export interface Vehicle {
  id: string;
  make: string;
  model: string;
  year: number;
  color: string;
  license_plate?: string;
  category: string;
  class_type?: string;
  transmission?: string;
  fuel_type?: string;
  daily_rate: number;
  is_available: boolean;
  image_filename?: string;
  image_url?: string;
  is_smoking_allowed?: boolean;
  store_id?: string;
  created_at?: string;
  updated_at?: string;
}

export interface VehicleListResponse {
  id: string;
  make: string;
  model: string;
  year: number;
  category: string;
  daily_rate: number;
  is_available: boolean;
  image_filename?: string;
  image_url?: string;
  transmission?: string;
  is_smoking_allowed?: boolean;
  store_id?: string;
}

export interface VehicleSearchParams {
  category?: string;
  make?: string;
  fuel_type?: string;
  is_available?: boolean;
  min_price?: number;
  max_price?: number;
  skip?: number;
  limit?: number;
}

export interface VehicleCreate {
  make: string;
  model: string;
  year: number;
  color: string;
  license_plate: string;
  category: string;
  class_type: string;
  transmission: string;
  fuel_type: string;
  daily_rate: number;
  image_filename?: string;
}

export interface VehicleUpdate {
  make?: string;
  model?: string;
  year?: number;
  color?: string;
  license_plate?: string;
  category?: string;
  class_type?: string;
  transmission?: string;
  fuel_type?: string;
  daily_rate?: number;
  is_available?: boolean;
  is_smoking_allowed?: boolean;
  image_filename?: string;
}

// 車両カテゴリの定数
export const VEHICLE_CATEGORIES = {
  compact: "コンパクト",
  standard: "スタンダード", 
  suv: "SUV",
  premium: "プレミアム",
  sports: "スポーツ",
  electric: "電気自動車",
  electric_suv: "電気SUV",
  convertible: "コンバーチブル",
  exotic: "エキゾチック",
  van: "バン・商用車"
} as const;

// 燃料タイプの定数
export const FUEL_TYPES = {
  gasoline: "ガソリン",
  electric: "電気",
  hybrid: "ハイブリッド",
  diesel: "ディーゼル"
} as const;

// トランスミッション
export const TRANSMISSIONS = {
  automatic: "オートマチック",
  manual: "マニュアル",
  cvt: "CVT"
} as const;

export type VehicleCategory = keyof typeof VEHICLE_CATEGORIES;
export type FuelType = keyof typeof FUEL_TYPES;
export type Transmission = keyof typeof TRANSMISSIONS;
